### PR TITLE
[SPARK-14936][BUILD][TESTS] FlumePollingStreamSuite is slow

### PR DESCRIPTION
**Pull request by keypointt**
https://issues.apache.org/jira/browse/SPARK-14936
## What changes were proposed in this pull request?

FlumePollingStreamSuite contains two tests which run for a minute each. This seems excessively slow and we should speed it up if possible.

In this PR, instead of creating `StreamingContext` directly from `conf`, here an underlying `SparkContext` is created before all and it is used to create  each`StreamingContext`. 

Running time is reduced by avoiding multiple `SparkContext` creations and destroys.
## How was this patch tested?

Tested on my local machine running `testOnly *.FlumePollingStreamSuite`

### CODE REVIEW

**Comment by zsxwing** [comment]
It still needs 1 minute each:

```
[info] FlumePollingStreamSuite:
[info] - flume polling test (1 minute, 8 seconds)
[info] - flume polling test multiple hosts (1 minute, 6 seconds)
```

**Comment by keypointt** [comment]
```
[info] FlumePollingStreamSuite: 
[info] - flume polling test (10 seconds, 396 milliseconds) 
[info] - flume polling test multiple hosts (10 seconds, 524 milliseconds)
```
It's faster this time, but I'm not sure this is the correct way to do it. 

I just keep underlying sparkcontext alive during streamingcontext start/stop, and finally stop the sparkcontext.

**Comment by keypointt** [comment]
Hi @zsxwing do you mind to review the changes here? thanks a lot

**Comment by zsxwing** [comment]
@keypointt you should create `SparkContext` in `beforeAll` and stop it in `afterAll`. Then use this `SparkContext` to create `StreamingContext`.

**Comment by zsxwing** [comment]
Could you revert other changes? (the typo fix is fine) They don't improve much but may make the tests flaky.

**Comment by keypointt** [comment]
sure I'll revert it, and add `beforeAll` / `afterAll `

File: external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumePollingStreamSuite.scala [hunk 1]
```diff
@@ -24,29 +24,41 @@ import scala.collection.JavaConverters._
...
 
-class FlumePollingStreamSuite extends SparkFunSuite with BeforeAndAfter with Logging {
+class FlumePollingStreamSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
 
   val maxAttempts = 5
   val batchDuration = Seconds(1)
 
+  @transient private var _sc: SparkContext = _
+
+  def sc: SparkContext = _sc
```
Comment (zsxwing) [inline external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumePollingStreamSuite.scala]: nit: you don't need this method.

File: external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumePollingStreamSuite.scala [hunk 2]
```diff
@@ -24,29 +24,41 @@ import scala.collection.JavaConverters._
...
     .set("spark.streaming.clock", "org.apache.spark.util.ManualClock")
 
   val utils = new PollingFlumeTestUtils
 
+  override def beforeAll(): Unit = {
+    _sc = new SparkContext(conf)
+  }
+
+  override def afterAll(): Unit = {
+    _sc.stop()
```
Comment (zsxwing) [inline external/flume/src/test/scala/org/apache/spark/streaming/flume/FlumePollingStreamSuite.scala]: nit: it's better to check `_sc != null`:

```
if (_sc != null) {
  _sc.stop()
  _sc = null
}
```
 So if `new SparkContext` fails, we don't hide it by a NPE.

**Comment by zsxwing** [comment]
By the way, could you update the description of this PR as well?

**Comment by keypointt** [comment]
I've updated the description, and changed the code as you recommended. Let's wait for the Jenkins result.

**Comment by zsxwing** [comment]
LGTM. Pending tests.

**Comment by zsxwing** [comment]
Merging to master and 2.0. Thanks, @keypointt

**Comment by keypointt** [comment]
@zsxwing thanks a lot for reviewing 👍

**Pull request closed without merging**